### PR TITLE
Remove quantity field from checkout model

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to remove quantity field from checkout model, and calculate it on the fly.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/checkout/utils.py [hunk 1]
```diff
@@ -651,3 +652,27 @@ def is_shipping_required(lines: Iterable["CheckoutLineInfo"]):
     return any(
         line_info.product.product_type.is_shipping_required for line_info in lines
     )
+
+
+def validate_variants_in_checkout_lines(lines):
```
Comment (fowczarek) [inline saleor/checkout/utils.py:657]: Maybe we could add typing :D :police_officer:

File: saleor/graphql/product/bulk_mutations/products.py [hunk 2]
```diff
@@ -127,6 +128,9 @@ def perform_mutation(cls, _root, info, ids, **data):
         if orders_id:
             recalculate_orders_task.delay(list(orders_id))
 
+        # delete checkout lines for deleted variants
+        checkout_models.CheckoutLine.objects.filter(variant_id__in=variants_ids)
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:132]: I think we have a lack of `@transaction.atomic` on this method.

Comment (fowczarek) [reply saleor/graphql/product/bulk_mutations/products.py:132]: We also should update the `quantity` field on the `Checkout` model.

File: saleor/graphql/product/bulk_mutations/products.py [hunk 3]
```diff
@@ -546,6 +550,9 @@ def perform_mutation(cls, _root, info, ids, **data):
         if orders_id:
             recalculate_orders_task.delay(list(orders_id))
 
+        # delete checkout lines for deleted variants
+        checkout_models.CheckoutLine.objects.filter(variant_id__in=pks)
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:554]: We also should update the `quantity` field on the `Checkout` model.

File: saleor/graphql/product/mutations/channels.py [hunk 4]
```diff
@@ -234,6 +235,9 @@ def remove_variants(
             channel=channel, variant__product_id=product
         ).exists():
             product_channel_listing.delete()
+        CheckoutLine.objects.filter(
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/channels.py:238]: We also should update the `quantity` field on the `Checkout` model.

File: saleor/graphql/product/mutations/products.py [hunk 5]
```diff
@@ -739,6 +740,10 @@ def perform_mutation(cls, _root, info, **data):
         order_models.OrderLine.objects.filter(pk__in=line_pks).delete()
         if orders_id:
             recalculate_orders_task.delay(list(orders_id))
+
+        # delete checkout lines for deleted variants
+        checkout_models.CheckoutLine.objects.filter(variant__id__in=variants_id)
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:745]: We also should update the `quantity` field on the `Checkout` model.

File: saleor/graphql/product/mutations/products.py [hunk 6]
```diff
@@ -1056,6 +1061,9 @@ def perform_mutation(cls, _root, info, **data):
         if orders_id:
             recalculate_orders_task.delay(list(orders_id))
 
+        # delete checkout lines for deleted variants
+        checkout_models.CheckoutLine.objects.filter(variant_id=instance.pk)
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:1065]: We also should update the `quantity` field on the `Checkout` model.

File: saleor/graphql/product/mutations/products.py [hunk 7]
```diff
@@ -739,6 +741,12 @@ def perform_mutation(cls, _root, info, **data):
         order_models.OrderLine.objects.filter(pk__in=line_pks).delete()
         if orders_id:
             recalculate_orders_task.delay(list(orders_id))
+
+        # update checkout quantity related to deleted lines
+        checkouts = checkout_models.Checkout.objects.filter(variant_id__in=variants_id)
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:746]: Can we figure out a different solution? This way has a big possibility to explode. We don't clean old checkout which it is easy to assume that we could have over 500k checkout objects. Assuming that 10% of this has the given variant assigned as a line will cause a huge amount of DB queries, `update_checkout_quantity` makes aggregation and save, so this action will cause around 100k queries. (Correct me if I am wrong) (this comment is related to all place where we call `update_checkout_quantity`)

Comment (fowczarek) [reply saleor/graphql/product/mutations/products.py:746]: Maybe we should rewrite `update_checkout_quantity` to use `F` query on the whole CheckoutQueryset?

Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:746]: I will try to find more optimised solution.

File: saleor/checkout/utils.py [hunk 8]
```diff
@@ -117,10 +119,13 @@ def add_variant_to_checkout(
         line.quantity = new_quantity
         line.save(update_fields=["quantity"])
 
-    checkout = update_checkout_quantity(checkout)
     return checkout
 
 
+def calculate_checkout_quantity(lines):
```
Comment (fowczarek) [inline saleor/checkout/utils.py:125]: Maybe?

```suggestion
def calculate_checkout_quantity(lines: Iterable["CheckoutLineInfo"]):
```

**Review by fowczarek — APPROVED** [review]

**Review by korycins — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Pull request merged**